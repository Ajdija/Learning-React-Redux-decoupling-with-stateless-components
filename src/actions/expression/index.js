import { DIGIT_CLICK_ACTION, CLICKED_DIGIT_KEY } from '../digit';

export const EXPRESSION_UPDATE_ACTION = 'EXPRESSION_UPDATE_ACTION';
export const NEW_EXPRESSION_KEY = 'NEW_EXPRESSION_KEY';

export const updateExpression = newExp => ({
  type: EXPRESSION_UPDATE_ACTION,
  payload: { [NEW_EXPRESSION_KEY]: newExp }
});

export const expressionReducer = (state = 0, { type, payload }) => {
  switch(type) {
    case EXPRESSION_UPDATE_ACTION:
      return (payload[NEW_EXPRESSION_KEY] || payload[NEW_EXPRESSION_KEY] === 0)
        ? payload[NEW_EXPRESSION_KEY]
        : state;
    case DIGIT_CLICK_ACTION:
      if(!state) {
        return payload[CLICKED_DIGIT_KEY];
      }
      return `${state}${payload[CLICKED_DIGIT_KEY]}`;
    default:
      return state;
  }
};
