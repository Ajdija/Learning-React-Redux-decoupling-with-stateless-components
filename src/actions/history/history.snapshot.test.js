import {
  historyReducer as reducer,
  updateHistory,
  UPDATE_HISTORY_ACTION
} from './index';
import { updateExpression, EXPRESSION_UPDATE_ACTION } from '../expression';

describe('history reducer', () => {
  it('should return initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it(`should handle ${EXPRESSION_UPDATE_ACTION} action`, () => {
    const action = updateExpression(10);
    const initialState = undefined;

    // First invoke
    const state1 = reducer(initialState, action);
    expect(state1).toMatchSnapshot();

    // Second invoke
    const action2 = updateExpression(15);
    const state2 = reducer(state1, action2);
    expect(state2).toMatchSnapshot();

    // Should not add the same item again
    const action3 = updateExpression(15);
    const state3 = reducer(state2, action3);
    expect(state3).toMatchSnapshot();
  });

  it(`should not add the same expression again when handling ${EXPRESSION_UPDATE_ACTION}`, () => {
    const action = updateExpression(10);
    const initialState = ["10"];

    const state1 = reducer(initialState, action);
    expect(state1).toMatchSnapshot();
  });

  it(`should handle ${UPDATE_HISTORY_ACTION} action`, () => {
    const action = updateExpression("1");
    const initialState = undefined;

    // First invoke
    const state1 = reducer(initialState, action);
    expect(state1).toMatchSnapshot();

    // Second invoke
    const action2 = updateExpression("5+5");
    const state2 = reducer(state1, action2);
    expect(state2).toMatchSnapshot();

    // Should not add the same item again
    const action3 = updateExpression("5+5");
    const state3 = reducer(state2, action3);
    expect(state3).toMatchSnapshot();
  });

  it(`should not add the same expression again when handling ${UPDATE_HISTORY_ACTION}`, () => {
    const action = updateExpression("5+5");
    const initialState = ["5+5"];

    const state1 = reducer(initialState, action);
    expect(state1).toMatchSnapshot();
  });
});
