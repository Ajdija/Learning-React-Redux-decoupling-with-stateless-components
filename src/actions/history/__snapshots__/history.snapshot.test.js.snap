// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`history reducer should handle EXPRESSION_UPDATE_ACTION action 1`] = `
Array [
  "10",
]
`;

exports[`history reducer should handle EXPRESSION_UPDATE_ACTION action 2`] = `
Array [
  "10",
  "15",
]
`;

exports[`history reducer should handle EXPRESSION_UPDATE_ACTION action 3`] = `
Array [
  "10",
  "15",
]
`;

exports[`history reducer should handle UPDATE_HISTORY_ACTION action 1`] = `
Array [
  "1",
]
`;

exports[`history reducer should handle UPDATE_HISTORY_ACTION action 2`] = `
Array [
  "1",
  "5+5",
]
`;

exports[`history reducer should handle UPDATE_HISTORY_ACTION action 3`] = `
Array [
  "1",
  "5+5",
]
`;

exports[`history reducer should not add the same expression again when handling EXPRESSION_UPDATE_ACTION 1`] = `
Array [
  "10",
]
`;

exports[`history reducer should not add the same expression again when handling UPDATE_HISTORY_ACTION 1`] = `
Array [
  "5+5",
]
`;

exports[`history reducer should return initial state 1`] = `Array []`;
